File: examples/download_media.js
#!/usr/bin/env node
// Download public Instagram photos & videos via the Instagram Cheapest API.
//
// Setup (Node 18+ — uses built-in fetch):
//   export RAPIDAPI_KEY='your-key'   // from https://rapidapi.com/liucccccccccccc/api/instagram-cheapest
//
// Usage:
//   node download_media.js Cabc123XYz                 // one post / Reel by shortcode
//   node download_media.js @nike                      // a whole profile
//   node download_media.js @nike --pages 5            // more pages of the profile
//   node download_media.js https://www.instagram.com/p/Cabc123XYz/   // a pasted URL works too
//
// Files are saved into ./downloads/. Instagram CDN links are signed and expire
// quickly, so this fetches the JSON and downloads immediately.
import { writeFile, mkdir } from "node:fs/promises";

const HOST = "instagram-cheapest.p.rapidapi.com";
const BASE_URL = `https://${HOST}/api/v1/instagram`;
const API_KEY = process.env.RAPIDAPI_KEY || "";

async function get(path, params = {}) {
  const qs = new URLSearchParams(params).toString();
  const res = await fetch(`${BASE_URL}/${path}${qs ? "?" + qs : ""}`, {
    headers: { "x-rapidapi-key": API_KEY, "x-rapidapi-host": HOST },
  });
  if (!res.ok) throw new Error(`HTTP ${res.status} for ${path}`);
  return res.json();
}

// Recursively collect best-resolution image/video URLs (handles carousels).
function findMedia(node, found = []) {
  if (Array.isArray(node)) {
    node.forEach((item) => findMedia(item, found));
  } else if (node && typeof node === "object") {
    const videos = node.video_versions;
    if (Array.isArray(videos) && videos.length) {
      const best = videos.reduce((a, b) => ((b.width || 0) > (a.width || 0) ? b : a));
      if (best.url) found.push({ type: "video", url: best.url });
    }
    const cands = node.image_versions2?.candidates;
    if (Array.isArray(cands) && cands.length && !(Array.isArray(videos) && videos.length)) {
      const best = cands.reduce((a, b) => ((b.width || 0) > (a.width || 0) ? b : a));
      if (best.url) found.push({ type: "photo", url: best.url });
    }
    Object.values(node).forEach((v) => findMedia(v, found));
  }
  return found;
}

async function downloadMedia(media, dir = "downloads", prefix = "ig") {
  await mkdir(dir, { recursive: true });
  for (let i = 0; i < media.length; i++) {
    const ext = media[i].type === "video" ? "mp4" : "jpg";
    const res = await fetch(media[i].url); // no API headers — public CDN
    const buf = Buffer.from(await res.arrayBuffer());
    const path = `${dir}/${prefix}_${i}.${ext}`;
    await writeFile(path, buf);
    console.log(`  saved ${path}`);
  }
}

async function downloadPost(code) {
  console.log(`Fetching post ${code} ...`);
  const post = await get("media_by_code2", { code });
  const media = findMedia(post);
  console.log(`Found ${media.length} media file(s)`);
  await downloadMedia(media, "downloads", code);
}

async function downloadProfile(username, maxPages = 3) {
  const profile = await get(`user/${username}`);
  const userId = profile.id || profile.pk;
  if (!userId) {
    console.log("Could not find user_id — inspect the raw profile JSON.");
    return;
  }
  console.log(`${username} -> user_id ${userId}`);
  let nextMaxId = null;
  for (let page = 0; page < maxPages; page++) {
    const params = { user_id: userId };
    if (nextMaxId) params.next_max_id = nextMaxId;
    const resp = await get("user_media", params);
    const media = findMedia(resp);
    console.log(`Page ${page + 1}: ${media.length} media file(s)`);
    await downloadMedia(media, `downloads/${username}`, `${username}_p${page}`);
    if (!resp.more_available) break;
    nextMaxId = resp.next_max_id;
    if (!nextMaxId) break;
    await new Promise((r) => setTimeout(r, 1000)); // ~1 req/sec on the free tier
  }
}

function parseTarget(target) {
  const t = target.trim().replace(/\/+$/, "");
  if (t.includes("instagram.com")) {
    const parts = t.split("/").filter(Boolean);
    for (const kw of ["p", "reel", "reels", "tv"]) {
      const i = parts.indexOf(kw);
      if (i !== -1) return ["post", parts[i + 1]];
    }
    return ["profile", parts[parts.length - 1].replace(/^@/, "")];
  }
  if (t.startsWith("@")) return ["profile", t.slice(1)];
  return ["post", t];
}

const argv = process.argv.slice(2);
if (!API_KEY) {
  console.error("Set RAPIDAPI_KEY first:  export RAPIDAPI_KEY='your-key'");
  process.exit(1);
}
if (!argv.length) {
  console.error("Usage: node download_media.js <shortcode | @username | url> [--pages N]");
  process.exit(1);
}
const pagesIdx = argv.indexOf("--pages");
const maxPages = pagesIdx !== -1 ? parseInt(argv[pagesIdx + 1], 10) : 3;
const target = argv.find((a, i) => !a.startsWith("--") && argv[i - 1] !== "--pages");

const [kind, value] = parseTarget(target);
if (kind === "profile") await downloadProfile(value, maxPages);
else await downloadPost(value);

File: examples/download_media.py
#!/usr/bin/env python3
"""Download public Instagram photos & videos via the Instagram Cheapest API.

Setup:
    pip install requests
    export RAPIDAPI_KEY='your-key'      # from https://rapidapi.com/liucccccccccccc/api/instagram-cheapest

Usage:
    python3 download_media.py Cabc123XYz                 # one post / Reel by shortcode
    python3 download_media.py @nike                      # a whole profile
    python3 download_media.py @nike --pages 5            # more pages of the profile
    python3 download_media.py https://www.instagram.com/p/Cabc123XYz/   # a pasted URL works too

Files are saved into ./downloads/. Note: Instagram CDN links are signed and
expire quickly, so this fetches the JSON and downloads immediately.
"""
import os
import sys
import json
import time
import argparse
import requests

HOST = "instagram-cheapest.p.rapidapi.com"
BASE_URL = f"https://{HOST}/api/v1/instagram"
API_KEY = os.environ.get("RAPIDAPI_KEY", "")
HEADERS = {"x-rapidapi-key": API_KEY, "x-rapidapi-host": HOST}


def get(path, params=None):
    """GET against the API (returns raw Instagram JSON)."""
    resp = requests.get(f"{BASE_URL}/{path}", headers=HEADERS, params=params, timeout=30)
    resp.raise_for_status()
    return resp.json()


def find_media(node, found=None):
    """Recursively collect best-resolution image/video URLs from the JSON.

    Handles single photos, videos/Reels, and carousels alike. If Instagram ever
    changes its response shape, print the JSON and verify these keys.
    """
    if found is None:
        found = []
    if isinstance(node, dict):
        videos = node.get("video_versions")
        if isinstance(videos, list) and videos:
            best = max(videos, key=lambda v: v.get("width", 0) or 0)
            if best.get("url"):
                found.append({"type": "video", "url": best["url"]})
        iv = node.get("image_versions2")
        if isinstance(iv, dict):
            candidates = iv.get("candidates") or []
            # Skip the poster image that videos also carry.
            if candidates and not (isinstance(videos, list) and videos):
                best = max(candidates, key=lambda c: c.get("width", 0) or 0)
                if best.get("url"):
                    found.append({"type": "photo", "url": best["url"]})
        for value in node.values():
            find_media(value, found)
    elif isinstance(node, list):
        for item in node:
            find_media(item, found)
    return found


def download_media(media, out_dir="downloads", prefix="ig"):
    """Stream each CDN URL to disk. No API key needed — these are public CDN links."""
    os.makedirs(out_dir, exist_ok=True)
    saved = []
    for i, m in enumerate(media):
        ext = "mp4" if m["type"] == "video" else "jpg"
        path = os.path.join(out_dir, f"{prefix}_{i}.{ext}")
        with requests.get(m["url"], timeout=60, stream=True) as r:
            r.raise_for_status()
            with open(path, "wb") as f:
                for chunk in r.iter_content(chunk_size=8192):
                    f.write(chunk)
        print(f"  saved {path}")
        saved.append(path)
    return saved


def download_post(code):
    print(f"Fetching post {code} ...")
    post = get("media_by_code2", params={"code": code})
    media = find_media(post)
    print(f"Found {len(media)} media file(s)")
    return download_media(media, prefix=code)


def download_profile(username, max_pages=3):
    profile = get(f"user/{username}")
    user_id = profile.get("id") or profile.get("pk")
    if not user_id:
        print("Could not find user_id. Raw profile (first 1 KB):")
        print(json.dumps(profile, indent=2)[:1000])
        return []
    print(f"{username} -> user_id {user_id}")
    saved, next_max_id, page = [], None, 0
    while page < max_pages:
        params = {"user_id": user_id}
        if next_max_id:
            params["next_max_id"] = next_max_id
        resp = get("user_media", params=params)
        media = find_media(resp)
        print(f"Page {page + 1}: {len(media)} media file(s)")
        saved += download_media(media, out_dir=f"downloads/{username}",
                                prefix=f"{username}_p{page}")
        if not resp.get("more_available"):
            break
        next_max_id = resp.get("next_max_id")
        if not next_max_id:
            break
        page += 1
        time.sleep(1)  # stay under the free tier's ~1 req/sec
    return saved


def parse_target(target):
    """Return ('post', shortcode) or ('profile', username) from any input."""
    t = target.strip().rstrip("/")
    if "instagram.com" in t:
        parts = [p for p in t.split("/") if p]
        for kw in ("p", "reel", "reels", "tv"):
            if kw in parts:
                return "post", parts[parts.index(kw) + 1]
        return "profile", parts[-1].lstrip("@")
    if t.startswith("@"):
        return "profile", t[1:]
    return "post", t


def main():
    parser = argparse.ArgumentParser(description="Download public Instagram media.")
    parser.add_argument("target", help="a shortcode, @username, or full instagram.com URL")
    parser.add_argument("--pages", type=int, default=3,
                        help="max pages when downloading a profile (default 3)")
    args = parser.parse_args()

    if not API_KEY:
        sys.exit("Set RAPIDAPI_KEY first:  export RAPIDAPI_KEY='your-key'")

    kind, value = parse_target(args.target)
    if kind == "profile":
        download_profile(value, max_pages=args.pages)
    else:
        download_post(value)


if __name__ == "__main__":
    main()
